lib/rspec/core/configuration.rb

```diff
@@ -1855,8 +1855,18 @@ def when_first_matching_example_defined(*filters)
 
       # @private
       def apply_derived_metadata_to(metadata)
-        @derived_metadata_blocks.items_for(metadata).each do |block|
-          block.call(metadata)
+        already_run_blocks = Set.new
+
+        # We loop and attempt to re-apply metadata blocks to support cascades
+        # (e.g. where a derived bit of metadata triggers the application of
+        # another piece of derived metadata, etc)
+        loop do
```

Comment:
Isn't there potential here for an infinite loop? Can we add a "break" that limits it to something even if it's rather high?

Reply:
As long as the user is not defining new `derived_metadata_blocks` within a derived metadata block, I think it is guaranteed to terminate.  Here's my reasoning:

* Assume that `@derived_metadata_blocks`  is not changing (I'll discuss that case separately).
* On a given loop iteration, one of two things will happen:
  * `@derived_metadata_blocks.items_for(metadata)` will return 1 or more blocks that have not previously been returned on a prior loop iteration
  * `@derived_metadata_blocks.items_for(metadata)` will return the same set of block that have been previously returned
* If it's the latter case, the `all?` will return true (because `already_run_blocks.include?(block)` will be true for all blocks), and the loop will break.
* If it's the former case, the newly returned blocks will be called and added to `already_run_blocks`, and the loop will iterate again.  However, we will be closer to having run _all_ of the blocks in `@derived_metadata_blocks`, and there is no way for the former case to happen indefinitely. At some point the latter case much occur because `@derived_metadata_blocks` has a finite number of blocks.

The assumption I stated above could be violated with some weird code like:

``` ruby
RSpec.configure do |c|
  c.define_derived_metadata(a: 0) do |meta|
    met[:a] += 1
    c.define_derived_metadata(a: meta[:a] + 1) do
      # ...
    end
  end
end
```

e.g. where the user is adding new blocks each time the block is triggered, and updating the metadata to match it.  I don't think this is a case any user will actually try, though, and if they do, well it's no different than a user writing a `let` that is self-referential and therefore recurses indefinitely.  Infinite recursion is possible with many constructs in Ruby and RSpec, and I don't think there's a viable way for us to guard against that without also preventing potentially valid use cases.

We could add a hard limit, but any limit would be arbitrary, and for any limit we can come up with, a valid use case exists for it happening once more.  Plus, it would be very odd to abort at an arbitrary point; it would be like a ruby method that recurses on itself arbitrarily stopping after some number of iterations.

Reply:
> it would be like a ruby method that recurses on itself arbitrarily stopping after some number of iterations.

Which is exactly what a `StackLevelTooDeep` error is... and this would hang indefinitely instead

Reply:
Good point; raising `SystemStackError` is reasonable.  I thought you were suggesting that we silently stop after a certain number of iterations.

I've updated this to raise `SystemStackError` after 200 iterations.

Reply:
Apologies for the confusion, I was advocating failing noisily if a limit was hit, I agree a stack error seems appropriate. I ponder if this could be refactored to be a recursive method to trigger the error organically at the same limit as the Ruby vm? Not a blocker however.

Reply:
I wondered the same thing, but can't think of a way to refactor it that way.

Reply:
Can you stamp this, @jonrowe?